common/src/main/java/io/druid/concurrent/LifecycleLock.java

```diff
@@ -0,0 +1,263 @@
+ *
+ * Example:
+ * class ExampleLifecycledClass
+ * {
+ *   final LifecycleLock lifecycleLock = new LifecycleLock();
+ *
+ *   void start()
+ *   {
+ *     if (!lifecycleLock.canStart()) {
+ *       .. return or throw exception
+ *     }
+ *     try {
+ *       .. do start
+ *       lifecycleLock.started();
+ *     }
+ *     finally {
+ *       lifecycleLock.exitStart();
+ *     }
+ *   }
+ *
+ *   void otherMethod()
+ *   {
+ *     Preconditions.checkState(lifecycleLock.awaitStarted());
+ *     // all actions done in start() are visible here
+ *     .. do stuff
+ *   }
+ *
+ *   void stop()
+ *   {
+ *     if (!lifecycleLock.canStop()) {
+ *       .. return or throw exception
+ *     }
+ *     // all actions done in start() are visible here
+ *     .. do stop
+ *   }
+ * }
+ */
+public final class LifecycleLock
+{
+  private static class Sync extends AbstractQueuedSynchronizer
+  {
+    private static final int NOT_STARTED = 0;
+    private static final int STARTING = 1;
+    private static final int STARTED = 2;
+    private static final int START_EXITED_SUCCESSFUL = 3;
+    private static final int START_EXITED_FAIL = 4;
+    private static final int STOPPING = 5;
+    private static final int STOPPED = 6;
+
+    boolean canStart()
+    {
+      return compareAndSetState(NOT_STARTED, STARTING);
+    }
+
+    void started()
+    {
+      if (!compareAndSetState(STARTING, STARTED)) {
+        throw new IllegalMonitorStateException("Called started() not in the context of start()");
+      }
+    }
+
+    void exitStart()
+    {
+      // see tryReleaseShared()
+      releaseShared(1);
+    }
+
+    @Override
+    protected boolean tryReleaseShared(int ignore)
+    {
+      while (true) {
+        int state = getState();
+        if (state == STARTING) {
+          if (compareAndSetState(STARTING, START_EXITED_FAIL)) {
+            return true;
+          }
+        } else if (state == STARTED) {
+          if (compareAndSetState(STARTED, START_EXITED_SUCCESSFUL)) {
+            return true;
+          }
+        } else {
+          throw new IllegalMonitorStateException("exitStart() called not in the end of the start() method");
+        }
+      }
+    }
+
+    boolean awaitStarted()
+    {
+      try {
+        // see tryAcquireShared()
+        acquireSharedInterruptibly(1);
+      }
+      catch (InterruptedException e) {
+        throw new RuntimeException(e);
+      }
+      return getState() == START_EXITED_SUCCESSFUL;
+    }
+
+    boolean awaitStarted(long timeNanos)
+    {
+      try {
+        // see tryAcquireShared()
+        if (!tryAcquireSharedNanos(1, timeNanos)) {
+          return false;
+        }
+      }
+      catch (InterruptedException e) {
+        throw new RuntimeException(e);
+      }
+      return getState() == START_EXITED_SUCCESSFUL;
+    }
+
+    @Override
+    protected int tryAcquireShared(int ignore)
+    {
+      return getState() >= STARTED ? 1 : -1;
```

Comment:
Should this be `getState() > STARTED`? It looks like with `>=`, awaitStarted would be able to proceed to the state check when the state is not yet some kind of `START_EXITED_*`, and so it might fail when it should have succeeded.